File: solutions/go/bank-account/2/bank_account.go
package account

import "sync"

// Define the Account type here.
type Account struct {
	amount int64
	closed bool
	mutext sync.Mutex
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}
	return &Account{amount: amount}
}

func (a *Account) Balance() (int64, bool) {
	a.mutext.Lock()
	defer a.mutext.Unlock()
	balance, closed := a.amount, a.closed
	return balance, !closed
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.mutext.Lock()
	defer a.mutext.Unlock()
	total, closed := a.amount+amount, a.closed
	if a.closed || total < 0 {
		return 0, false
	}
	a.amount = total
	return total, !closed
}

func (a *Account) Close() (int64, bool) {
	a.mutext.Lock()
	defer a.mutext.Unlock()
	total, closed := a.amount, a.closed
	a.closed = true
	a.amount = 0
	return total, !closed
}

File: solutions/csharp/bank-account/2/BankAccount.cs
public class BankAccount
{
    private decimal balance;
    private bool isOpen;
    private readonly object balance_lock = new();

    public void Open()
    {
        if (isOpen)
            throw new InvalidOperationException("Account is already open.");
        lock (balance_lock)
            isOpen = true;
    }

    public void Close()
    {
        EnsureAccountIsOpen();
        lock (balance_lock)
        {
            balance = 0;
            isOpen = false;
        }
    }

    public decimal Balance
    {
        get
        {
            EnsureAccountIsOpen();
            return balance;
        }
    }

    public void Deposit(decimal change)
    {
        EnsureAccountIsOpen();
        if (change <= 0)
            throw new InvalidOperationException("Deposit amount must be positive.");
        lock (balance_lock)
            balance += change;
    }

    public void Withdraw(decimal change)
    {
        EnsureAccountIsOpen();
        if (change > balance || change <= 0)
            throw new InvalidOperationException("Change must be higher than 0 and lower than current balance.");
        lock (balance_lock)
            balance -= change;
    }

    private void EnsureAccountIsOpen()
    {
        if (!isOpen)
            throw new InvalidOperationException("Account is closed.");
    }
}
